types/src/lib.rs

```diff
@@ -0,0 +1,65 @@
+//! Fundamental types shared by other Propolis crates.
+//!
+//! This crate defines some basic types that are shared by multiple other
+//! Propolis crates (library, client, server, and/or standalone) such that they
+//! can all use those types (and implement their own conversions to/from them)
+//! without any layering oddities.
+
+use std::io::ErrorKind;
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+/// A PCI bus/device/function tuple. Supports conversion from a string formatted
+/// as "B.D.F", e.g. "0.7.0".
+#[derive(
+    Clone, Copy, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord, Debug,
+)]
+pub struct PciPath(pub u8, pub u8, pub u8);
```

Comment:
Unless we're going to further validate this, then exposing the latter two fields publicly means allowing invalid BDFs (since function must be 0-7 and device/slice must be 0-31)

Reply:
These should still be validated by the Propolis lib/server, in the sense that the lib only accepts `pci::Bdf`s and the conversion from `PciPath` to `Bdf` is fallible and will reject invalid device/function numbers. But it's not too hard to just reject invalid paths up front, and IMO doing that carries some other advantages, so I tightened up the bolts as follows:

- This struct's fields are now private, and it has a constructor that verifies that the device and function are in range.
- It also now serializes as "B.D.F" instead of "PciPath(B, D, F)"; this lets the deserializer just forward to `from_str` and settles on "B.D.F" as the One True Way to represent a string-format BDF. (I also added some serialization unit tests.)